ibis/backends/clickhouse/__init__.py

```diff
@@ -114,3 +128,102 @@ def list_tables(self, like=None, database=None):
         databases = list(data[0])
         return self._filter_with_like(databases, like)
+
+    def raw_sql(self, query: str, external_tables={}):
+        external_tables_list = []
+        for name, df in external_tables.items():
+            if not isinstance(df, pd.DataFrame):
+                raise TypeError(
+                    'External table is not an instance of pandas ' 'dataframe'
+                )
+            schema = sch.infer(df)
+            external_tables_list.append(
+                {
+                    'name': name,
+                    'data': df.to_dict('records'),
+                    'structure': list(
+                        zip(
+                            schema.names,
+                            [
+                                str(ClickhouseDataType.from_ibis(t))
+                                for t in schema.types
+                            ],
+                        )
+                    ),
+                }
+            )
+
+        ibis.util.log(query)
+        return self.con.execute(
+            query,
+            columnar=True,
+            with_column_types=True,
+            external_tables=external_tables_list,
+        )
+
+    def ast_schema(self, query_ast, external_tables={}):
+        # Allowing signature to accept `external_tables`
+        return super().ast_schema(query_ast)
+
+    def fetch_from_cursor(self, cursor, schema):
+        data, columns = cursor
+        if not len(data):
+            # handle empty resultset
+            return pd.DataFrame([], columns=schema.names)
+
+        df = pd.DataFrame.from_dict(OrderedDict(zip(schema.names, data)))
+        return schema.apply_to(df)
+
+    def close(self):
+        """Close Clickhouse connection and drop any temporary objects"""
+        self.con.disconnect()
+
+    def _fully_qualified_name(self, name, database):
+        if bool(fully_qualified_re.search(name)):
+            return name
+
+        database = database or self.current_database
+        return f'{database}.`{name}`'
+
+    def get_schema(self, table_name, database=None):
+        """
+        Return a Schema object for the indicated table and database
+
+        Parameters
+        ----------
+        table_name : string
+          May be fully qualified
+        database : string, default None
+
+        Returns
+        -------
+        schema : ibis Schema
+        """
+        qualified_name = self._fully_qualified_name(table_name, database)
+        query = f'DESC {qualified_name}'
+        data, columns = self.raw_sql(query)
+        return sch.schema(
+            data[0], list(map(ClickhouseDataType.parse, data[1]))
+        )
+
+    def set_options(self, options):
+        self.con.set_options(options)
+
+    def reset_options(self):
+        # Must nuke all cursors
+        raise NotImplementedError
+
+    def _ensure_temp_db_exists(self):
+        name = (options.clickhouse.temp_db,)
+        if name not in self.list_databases():
```

Comment:
Is `list_databases` returning a `list[tuple[str]]`? That seems like it violates the exist contract. Is this just a carryover from the move?